src/helpers/transport/receive.rs

```diff
@@ -1,51 +1,74 @@
 use crate::helpers::transport::stream::{StreamCollection, StreamKey};
 use futures::Stream;
 use futures_util::StreamExt;
-#[cfg(any(not(feature = "web-app"), feature = "test-fixture", test))]
-use std::convert::identity;
 use std::{
+    error::Error as StdError,
     pin::Pin,
     task::{Context, Poll},
 };
+use tracing::error;
 
-/// Represents a stream of records.
-/// If stream is not received yet, each poll generates a waker that is used internally to wake up
-/// the task when stream is received.
-/// Once stream is received, it is moved to this struct and it acts as a proxy to it.
-pub struct ReceiveRecords<S, T = S, F = fn(T) -> S>
+/// Adapt a stream of `Result<T: Into<Vec<u8>>, Error>` to a stream of `Vec<u8>`.
+///
+/// If an error is encountered, the error is logged, and the stream is terminated.
+pub struct LogErrors<S, T, E>
 where
-    F: FnOnce(T) -> S,
+    S: Stream<Item = Result<T, E>> + Unpin,
+    T: Into<Vec<u8>>,
+    E: StdError,
 {
-    inner: ReceiveRecordsInner<S, T, F>,
+    inner: S,
 }
 
-impl<S> ReceiveRecords<S, S, fn(S) -> S> {
-    #[cfg(any(not(feature = "web-app"), feature = "test-fixture", test))]
-    pub(crate) fn new(key: StreamKey, coll: StreamCollection<S>) -> Self {
-        Self {
-            inner: ReceiveRecordsInner::Pending(key, coll, Some(identity)),
-        }
+impl<S, T, E> LogErrors<S, T, E>
+where
+    S: Stream<Item = Result<T, E>> + Unpin,
+    T: Into<Vec<u8>>,
+    E: StdError,
+{
+    pub fn new(inner: S) -> Self {
+        Self { inner }
     }
 }
 
-impl<S, T, F> ReceiveRecords<S, T, F>
+impl<S, T, E> Stream for LogErrors<S, T, E>
 where
-    F: FnOnce(T) -> S,
+    S: Stream<Item = Result<T, E>> + Unpin,
+    T: Into<Vec<u8>>,
+    E: StdError,
 {
-    #[cfg(feature = "web-app")]
-    pub(crate) fn mapped(key: StreamKey, coll: StreamCollection<T>, f: F) -> Self {
+    type Item = Vec<u8>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        match Pin::get_mut(self).inner.poll_next_unpin(cx) {
+            Poll::Pending => Poll::Pending,
+            Poll::Ready(Some(Ok(chunk))) => Poll::Ready(Some(chunk.into())),
+            Poll::Ready(Some(Err(err))) => {
+                error!("error reading records: {err}");
```

Comment:
I wonder if we should panic in this case. This does not seem like a recoverable error and if we just return None, then our protocols eventually hang with no clear indication what went wrong

Reply:
Are we looking for a panic, or should this return `Poll<Option<Result<Vec<u8>>>>` (i.e., `type Item = Result<Vec<u8>>`), so that the read is fallible?  (Not here, but on ReceiveRecords.)

Reply:
that is another option, however at some point we need to unwrap this result and it does not seem that we can recover if some part of records stream is missing. so clients are no better positioned to decide because the best they can do is `unwrap` imo

Reply:
You're right - we could potentially make it work with panics and runtime-per-query approach where panic will just abort the the thread but it is worse than returning `Result` imo

Reply:
That outcome works for me, but can we comment to that effect here so that we don't lose this insight?